### PR TITLE
gh-120834: Use PyGenObject for generators, coroutines and async generators

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-120834

### CODE REVIEW

File: Include/cpython/genobject.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ PyAPI_FUNC(PyObject *) PyCoro_New(PyFrameObject *,
 
 /* --- Asynchronous Generators -------------------------------------------- */
 
-typedef struct _PyAsyncGenObject PyAsyncGenObject;
+typedef PyGenObject PyAsyncGenObject;
```
Comment (iritkatriel) [inline Include/cpython/genobject.h:38]: PyCoroObject and PyAsyncGenObject are mentioned in the docs, so I think we need to leave this even though we're not using these names.

File: Include/internal/pycore_genobject.h [hunk 2]
```diff
@@ -10,40 +10,25 @@ extern "C" {
...
-
 struct _PyGenObject {
-    /* The gi_ prefix is intended to remind of generator-iterator. */
-    _PyGenObject_HEAD(gi)
-};
-
-struct _PyCoroObject {
-    _PyGenObject_HEAD(cr)
+    /* The gi_ prefix is for generator-iterator. */
+    PyObject_HEAD                                               \
```
Comment (markshannon) [inline Include/internal/pycore_genobject.h:15]: We don't need the line extension `\`s any more.

File: Python/_warnings.c [hunk 3]
```diff
@@ -1399,12 +1399,12 @@ PyErr_WarnExplicitFormat(PyObject *category,
 }
 
 void
-_PyErr_WarnUnawaitedAgenMethod(PyAsyncGenObject *agen, PyObject *method)
+_PyErr_WarnUnawaitedAgenMethod(PyGenObject *agen, PyObject *method)
```
Comment (markshannon) [inline Python/_warnings.c:1402]: This seems a less correct type. `agen` should be an async generator, so `PyAsyncGenObject *` seems the correct type.

File: Python/_warnings.c [hunk 4]
```diff
@@ -1413,7 +1413,7 @@ _PyErr_WarnUnawaitedAgenMethod(PyAsyncGenObject *agen, PyObject *method)
 
 
 void
-_PyErr_WarnUnawaitedCoroutine(PyObject *coro)
+_PyErr_WarnUnawaitedCoroutine(PyGenObject *coro)
```
Comment (markshannon) [inline Python/_warnings.c:1416]: This should only be called with a coroutine, so I think `PyCoroObject *` is correct type.
Maybe add an assert that `Py_TYPE(coro) == &PyCoro_Type`?

File: Include/internal/pycore_ceval.h [hunk 5]
```diff
@@ -232,7 +232,7 @@ static inline void _Py_LeaveRecursiveCall(void)  {
 
 extern struct _PyInterpreterFrame* _PyEval_GetFrame(void);
 
-PyAPI_FUNC(PyObject *)_Py_MakeCoro(PyFunctionObject *func);
+PyAPI_FUNC(PyGenObject *)_Py_MakeCoro(PyFunctionObject *func);
```
Comment (markshannon) [inline Include/internal/pycore_ceval.h:235]: We generally have a convention that there is a correspondence between the C type and the Python class.
e.g. if `PyXXX_Check[Exact](obj)` is true, then it is safe to cast `(PyXXXObject *)obj` and vice-versa.
This sort of breaks that for coroutines and async generators.

So, I think it best to leave this as `PyObject *`.  Does that make sense to you?

**Review by markshannon — CHANGES_REQUESTED** [review]
A few pedantic type suggestions, otherwise LGTM.

File: Python/_warnings.c (hunk 3 above)
Comment (iritkatriel) [reply Python/_warnings.c:1402]: Are we going to keep PyAsyncGenObject and PyCoroObject, now that they are just aliases?
It's not like using them in a signature will give compile-time type checks.

I was assuming we will want to stop using them altogether and deprecate them from the API.

**Comment by iritkatriel** [comment]
We decided not to do this after all.

**Pull request closed without merging**